packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts

```diff
@@ -0,0 +1,228 @@
+import { toInstrumentedOdspTokenFetcher } from "./odspDocumentServiceFactoryCore";
+import { evalBlobsAndTrees, ISnapshotCacheValue } from "./odspDocumentStorageManager";
+import {
+    createOdspLogger,
+    fetchAndParseAsJSONHelper,
+    getOdspResolvedUrl,
+    getWithRetryForTokenRefresh,
+    IOdspResponse,
+} from "./odspUtils";
+import { IOdspSnapshot } from "./contracts";
+
+/**
+ * Function to prefetch the snapshot and cached it in the persistant cache, so that when the container is loaded
+ * the cached latest snapshot could be used and removes the network call from the critical path.
+ * @param resolvedUrl - Resolved url to fetch the snapshot.
+ * @param getStorageToken - function that can provide the storage token for a given site. This is
+ *  is also referred to as the "VROOM" token in SPO.
+ * @param persistedCache - Cache to store the fetched snapshot.
+ * @param logger - Logger to have telemetry events.
+ * @param hostSnapshotFetchOptions - Options to fetch the snapshot if any. Otherwise default will be used.
+ * @returns - True if the snapshot is cached, false otherwise.
+ */
+export async function prefetchLatestSnapshot(
+    resolvedUrl: IResolvedUrl,
+    getStorageToken: TokenFetcher<OdspResourceTokenFetchOptions>,
+    persistedCache: IPersistedCache = new LocalPersistentCache(),
+    logger: ITelemetryBaseLogger,
+    hostSnapshotFetchOptions: ISnapshotOptions | undefined,
+): Promise<boolean> {
+    const odspLogger = createOdspLogger(ChildLogger.create(logger, "PrefetchSnapshot"));
+    const odspResolvedUrl = getOdspResolvedUrl(resolvedUrl);
+    const snapshotUrl = odspResolvedUrl.endpoints.snapshotStorageUrl;
+    let snapshotOptions: ISnapshotOptions = {
+        deltas: 1,
+        channels: 1,
+        blobs: 2,
+    };
+
+    if (hostSnapshotFetchOptions) {
+        snapshotOptions = {...snapshotOptions, ...hostSnapshotFetchOptions};
+    }
+
+    const storageTokenFetcher = toInstrumentedOdspTokenFetcher(
+        odspLogger,
+        odspResolvedUrl,
+        getStorageToken,
+        true /* throwOnNullToken */,
+    );
+
+    const snapshotCacheEntry: ICacheEntry = {
+        type: snapshotKey,
+        key: "",
+        file: {
+            resolvedUrl: odspResolvedUrl,
+            docId: odspResolvedUrl.hashedDocumentId,
+        },
+    };
+
+    return getWithRetryForTokenRefresh(async (tokenFetchOptions) => {
+        // No need to ask cache twice - if first request was unsuccessful, cache unlikely to have data on second turn.
+        if (tokenFetchOptions.refresh) {
+            await fetchSnapshot(
+                snapshotUrl,
+                odspResolvedUrl,
+                snapshotOptions,
+                tokenFetchOptions,
+                storageTokenFetcher,
+                odspLogger,
+                persistedCache,
+                snapshotCacheEntry,
+            );
+        } else {
+            await PerformanceEvent.timedExecAsync(
```

Comment:
Curious, why not wrap whole getWithRetryForTokenRefresh callback? We are not logging errors on second call, which might be useful to log.
Also if you wrap whole thing, please record if it's second or first call (take a look at other places how we do it)

Reply:
Changed and noted the attempt number too.